Add basic report command with alert and event

Pugzy opened the pull request:
Adds a basic report command which takes a player and a reason. This triggers a `PlayerReportEvent` and sends a formatted message to all players with the permission `Permissions.ADMINCHAT`.

e.g. image below (both player names are clickable, ignore my messed up flairs)

![image](https://user-images.githubusercontent.com/8608892/71695923-7ee7cf00-2dab-11ea-85a8-34c61f92bd75.png)

Signed-off-by: Pugzy <pugzy@mail.com>

TheMolkaPL commented on src/main/java/tc/oc/pgm/commands/ReportCommands.java:
```diff
@@ -0,0 +1,51 @@
+                    + "{1}"
+                    + ChatColor.YELLOW
+                    + ": "
+                    + "{2}",
```
I think this should be defined in a constant?

Pugzy replied:
Moved to components rather than a string.

TheMolkaPL commented on src/main/java/tc/oc/pgm/events/PlayerReportEvent.java:
```diff
@@ -0,0 +1,37 @@
+import tc.oc.pgm.api.player.event.MatchPlayerEvent;
+
+/** Called immediately AFTER a player runs the report command. */
+public class PlayerReportEvent extends MatchPlayerEvent {
```
Should be cancelable.

TheMolkaPL commented on src/main/java/tc/oc/pgm/events/PlayerReportEvent.java:
```diff
@@ -0,0 +1,37 @@
+  public PlayerReportEvent(MatchPlayer sender, MatchPlayer player, String reason) {
+    super(player);
+    this.sender = sender;
+    this.reason = reason;
```
```java
this.sender = checkNotNull(sender);
this.reason = checkNotNull(reason);
```

TheMolkaPL commented on src/main/java/tc/oc/pgm/commands/ReportCommands.java:
```diff
@@ -0,0 +1,51 @@
+                accused.getStyledName(NameStyle.FANCY),
+                new PersonalizedText(reason.trim())));
+
+    match.getPlayers().stream()
```
Before sending to players please log this in a logger.

Electroid replied:
Sending to console is good for this

Pugzy replied:
Now sending to console with `[A]` prefix removed.

TheMolkaPL commented:
Imo there should also be `/reports` which prints last reports. In case all admins were offline and you want to see who got reported during that time.

Brottweiler commented:
> Imo there should also be `/reports` which prints last reports. In case all admins were offline and you want to see who got reported during that time.

I do not think we need to store reports, it isn't *that* useful for us. Simply display them to all online staff.

Pablete1234 commented:
The event is launched precisely to allow a 3rd party plugin persisting the data if they want to. I don't think PGM needs to store reports for them to be effective, rather treat the base /report as a way to message online staff.

Also, i'm not convinced that it should be prefixed with `[R]` instead of `[A] PlayerA reported PlayerB: reason`, for admin chat

Pablete1234 commented on src/main/java/tc/oc/pgm/commands/ReportCommands.java:
```diff
@@ -0,0 +1,51 @@
+import tc.oc.pgm.events.PlayerReportEvent;
+import tc.oc.util.components.Components;
+
+public class ReportCommands {
```
I think this would fit better as just ModerationCommands, for when other commands are added

Pugzy replied:
Renamed

Pablete1234 commented on src/main/java/tc/oc/pgm/commands/ReportCommands.java:
```diff
@@ -0,0 +1,51 @@
+                    + "{1}"
+                    + ChatColor.YELLOW
+                    + ": "
+                    + "{2}",
```
I'm pretty sure this would be shorter if you used the text component stuff instead of string concat:
```java
new PersonalizedComponent(
  new PersonalizedComponent("["),
  new PersonalizedComponent("A", ChatColor.GOLD),
  new PersonalizedComponent("] "),
  sender.getStyledName(NameStyle.FANCY),
  new PersonalizedComponent(" \u2794 ", ChatColor.YELLOW),
...
```

Pugzy replied:
Changed to components.

Pugzy commented:
New formatted screenshot: 
![image](https://user-images.githubusercontent.com/8608892/71771171-0b6eca80-2f2f-11ea-83b1-3e19629fb477.png)

How can I prevent console from running the command?
Should the sender receive a message if the event is cancelled?

TheMolkaPL commented on src/main/java/tc/oc/pgm/commands/ModerationCommands.java:
```diff
@@ -0,0 +1,55 @@
+    PlayerReportEvent event = new PlayerReportEvent(sender, accused, reason);
+    match.callEvent(event);
+
+    if (!event.isCancelled()) {
```
It would be cleaner with

```java
if (event.isCancelled()) return;
```

TheMolkaPL commented:
> How can I prevent console from running the command?

Why would you prevent?

> Should the sender receive a message if the event is cancelled?

Hmm, how about using `ExtendedCancellable` instead of the standard `Cancellable`?

Pugzy commented:
Replaced sender in the event to be a `CommandSender` to allow for Console reports.

Updated event to extend `ExtendedCancellable` and relayed message to the sender.

Pablete1234 commented on src/main/java/tc/oc/pgm/commands/ModerationCommands.java:
```diff
@@ -0,0 +1,70 @@
+      Match match,
+      Player player,
+      @Text String reason) {
+    MatchPlayer accused = match.getPlayer(player);
```
What will happen if the player is in another match? i don't think the Match should be a parameter for this command, you should make use of the MatchManager to get the MatchPlayer

Electroid replied:
I think the report should be scoped to the Match. The issue here is you need to check that `accused != null`

Pugzy replied:
> What will happen if the player is in another match?

We don't have the ability to run multiple matches at the moment and I don't think every new change needs to unnecessary pander to this idea. 

Regardless you shouldn't be reporting players from alternative matches because as you said yourself they should act independently. The audiences that it broadcast the notification message to could be server-wide as moderators may wish to switch match to investigate.

> you need to check that `accused != null`

When would the accused not be a player? The command framework seems pretty hell-bent on making sure the correct arguments are passed to the function.

![image](https://user-images.githubusercontent.com/8608892/71772520-31eb3080-2f44-11ea-92ca-ecc0699578d1.png)

Electroid commented on src/main/java/tc/oc/pgm/commands/ModerationCommands.java:
```diff
@@ -0,0 +1,70 @@
+
+    if (event.isCancelled()) {
+      if (event.getCancelMessage() != null) {
+        commandSender.sendMessage(event.getCancelMessage());
```
`matchPlayer.sendMessage`

Pugzy replied:
The use of `CommandSender` allows for console reports. I can remove this feature if someone lets me know how to prevent Console from running the command as requested previously in a discussion with Molka.

Electroid commented on src/main/java/tc/oc/pgm/commands/ModerationCommands.java:
```diff
@@ -0,0 +1,70 @@
+      return;
+    }
+
+    commandSender.sendMessage(
```
`matchPlayer.sendMessage`
